filters/shedder/admission.go

```diff
@@ -0,0 +1,366 @@
+		}
+	}
+
+	windowSize := defaultWindowSize
+	if len(args) > 3 {
+		windowSize, err = getIntArg(args[3])
+		if err != nil {
+			logrus.Warnf("windowsize failed: %v", err)
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		if windowSize < minWindowSize {
+			windowSize = minWindowSize
+		}
+		if windowSize > maxWindowSize {
+			windowSize = maxWindowSize
+		}
+	}
+
+	minRequests := defaultMinRps
+	if len(args) > 4 {
+		minRequests, err = getIntArg(args[4])
+		if err != nil {
+			logrus.Warnf("minRequests failed: %v", err)
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	threshold := defaultThreshold
+	if len(args) > 5 {
+		threshold, err = getFloat64Arg(args[5])
+		if err != nil {
+			logrus.Warnf("threshold failed %v", err)
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	maxRejectProbability := defaultMaxRejectProbability
+	if len(args) > 6 {
+		maxRejectProbability, err = getFloat64Arg(args[6])
+		if err != nil {
+			logrus.Warnf("maxRejectProbability failed: %v", err)
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	aggression := defaultAggression
+	if len(args) > 7 {
+		aggression, err = getFloat64Arg(args[7])
+		if err != nil {
+			logrus.Warnf("aggression failed: %v", err)
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		if aggression <= 0.0 {
+			logrus.Warn("aggression should be >0")
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	ac := &admissionControl{
+		metrics:      metrics.Default,
+		metricSuffix: metricSuffix,
+		tracer:       spec.tracer,
+
+		mode:                 mode,
+		d:                    d,
+		windowSize:           windowSize,
+		minRequests:          minRequests,
+		successThreshold:     threshold,
+		maxRejectProbability: maxRejectProbability,
+		aggression:           aggression,
+
+		totals:         make([]int64, windowSize),
+		success:        make([]int64, windowSize),
+		counter:        atomic.NewInt64(0),
+		successCounter: atomic.NewInt64(0),
+	}
+	go ac.tickWindows()
+	return ac, nil
+}
+
+func (ac *admissionControl) tickWindows() {
+	t := time.NewTicker(ac.d)
+	i := 0
+	for range t.C {
+		val := ac.counter.Swap(0)
+		ok := ac.successCounter.Swap(0)
+
+		ac.mu.Lock()
+		ac.totals[i] = val
+		ac.success[i] = ok
+		ac.mu.Unlock()
+
+		i = (i + 1) % ac.windowSize
+	}
+}
+
+func (ac *admissionControl) count() (float64, float64) {
+	ac.mu.Lock()
+	defer ac.mu.Unlock()
+	return float64(sum(ac.totals)), float64(sum(ac.success))
+}
+
+func sum(a []int64) int64 {
+	var result int64
+	for _, v := range a {
+		result += v
+	}
+	return result
+}
+
+// calculates P_{reject} see https://opensource.zalando.com/skipper/reference/filters/#admissioncontrol
+func (ac *admissionControl) pReject() float64 {
+	var rejectP float64
+
+	total, success := ac.count()
+	factor := float64(ac.d) / float64(time.Second)
+	if total < float64(ac.minRequests)*factor {
+		return -1
+	}
+
+	s := success / ac.successThreshold
```

Comment:
> Success threshold sets the lowest request success rate at which the
filter will not reject requests. It is within $(0,1]$ and defaults to
0.95, which means an error rate of lower than 5% will not trigger
rejects.

Maybe it would be cleaner if we calculate
```go
successRate := success / total
if successRate > ac.successThreshold {
    return -1
}
```

Reply:
s is reused in the calculation and check the envoy link to doc https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/admission_control_filter#overview